Watch destroyed Fragments (part 2)

pyricau opened the pull request:
This is a follow up for #894, introducing support for fragments again but without adding the support library dependency to LeakCanary.

- Added configuration for watching fragments directly on AndroidRefWatcherBuilder
- Moved ActivityRefWatcher to the internal package, it shouldn't have been an API. This is a breaking change, but should have a small impact (it's easy to recreate this class and tailor it to ones's needs)
- New `leakcanary-support-fragment` module to watch support fragments.
- Watching Fragments is enabled by default (can be disabled). If enabled, it'll automatically watch native Fragments on Android O and it'll watch support fragments if the `leakcanary-support-fragment` dependency is in the classpath.

Fixes #983

JakeWharton commented on leakcanary-android/src/main/java/com/squareup/leakcanary/internal/AndroidOFragmentRefWatcher.java:
```diff
@@ -0,0 +1,51 @@
+import android.os.Build;
+import com.squareup.leakcanary.RefWatcher;
+
+@TargetApi(Build.VERSION_CODES.O) //
```
Target API should never be used. You want requires API.

JakeWharton commented on leakcanary-support-fragment/build.gradle:
```diff
@@ -0,0 +1,25 @@
+  // However, FragmentLifecycleCallbacks wasn't properly usable until 25.2.0 where it became static
+  // https://developer.android.com/topic/libraries/support-library/rev-archive#fixed-issues_12
+  //noinspection GradleDependency
+  api 'com.android.support:support-fragment:25.2.0'
```
implementation

pyricau commented:
Added instrumentation tests to confirm we were detecting fragment leaks. Turns out we weren't, because activity destroy callback happens before fragment destroy and we stopped listening before we had a change to watch. I removed the unregistering in activity destroy since the fragment callbacks are automatically unregistered any way.

Yay for tests! 🎉

pyricau commented on leakcanary-android-instrumentation/src/androidTest/java/com/squareup/leakcanary/InstrumentationTestApplication.java:
```diff
@@ -5,9 +5,6 @@
     super.onCreate();
-    if (LeakCanary.isInAnalyzerProcess(this)) {
-      return;
-    }
```
Note: somewhat unrelated but I thought of this when writing the fragment tests. We're never running the leak canary process during instrumentation tests, so no need for this.

pyricau commented on gradle/gradle-mvn-push.gradle:
```diff
@@ -81,6 +81,7 @@ afterEvaluate { project ->
     if (plugins.hasPlugin('com.android.application') || plugins.hasPlugin('com.android.library')) {
       task androidJavadocs(type: Javadoc) {
         source = android.sourceSets.main.java.srcDirs
+        exclude "**/internal/**"
```
The javadoc talk failed because dependencies other than Android aren't in the javadoc classpath. This doesn't fix that, but only internal classes used code from external dependencies, and we shouldn't generate javadoc for internal classes anyway. We might still need to actually fix the javadoc task in the future though.

vanniktech replied:
You can add them back. https://github.com/vanniktech/gradle-maven-publish-plugin/blob/0b2c99e74f00da096a2a8a7ff27b51ead5009ac2/src/main/groovy/com/vanniktech/maven/publish/MavenPublishPlugin.groovy#L102

```
// Append also the classpath and files for release library variants. This fixes the javadoc warnings.
// Got it from here - https://github.com/novoda/bintray-release/pull/39/files
def releaseVariant = project.android.libraryVariants.toList().last()
classpath += releaseVariant.javaCompile.classpath
classpath += releaseVariant.javaCompile.outputs.files
```

pyricau replied:
nice!

pyricau replied:
🙏https://github.com/square/leakcanary/pull/1012

jrodbx replied:
Added comment relating to this snippet here: https://github.com/square/leakcanary/pull/1012#discussion_r196144602
